src/main/java/org/candlepin/subscriptions/security/AntiCsrfFilter.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.subscriptions.security;
+
+import org.candlepin.subscriptions.ApplicationProperties;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.annotation.Order;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Filter that prevents CSRF by verifying for a valid Origin or Referer header.
+ */
+@Component
+@Order(1)
+public class AntiCsrfFilter extends OncePerRequestFilter {
+
+    private static final List<String> MODIFYING_METHODS = Arrays.asList("POST", "PUT", "DELETE", "PATCH");
+    private static Logger log = LoggerFactory.getLogger(AntiCsrfFilter.class);
+
+    private final boolean disabled;
+
+    AntiCsrfFilter(ApplicationProperties props, ConfigurableEnvironment env) {
+        disabled = props.isDevMode() || Arrays.asList(env.getActiveProfiles()).contains("capacity-ingress");
+        if (disabled) {
+            log.info("Origin & Referer checking (anti-csrf) disabled.");
+        }
+        else {
+            log.info("Origin & Referer checking (anti-csrf) enabled.");
+        }
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
+        FilterChain filterChain) throws ServletException, IOException {
+        /* origin comes first as it is much faster to parse when present */
+        if (disabled || requestTypeOkay(request) || originMatches(request) || refererMatches(request)) {
+            filterChain.doFilter(request, response);
+        }
+        else {
+            response.sendError(403, "Origin & Referer both bad. Cross origin requests not allowed.");
+        }
+    }
+
+    private boolean requestTypeOkay(HttpServletRequest request) {
+        return !MODIFYING_METHODS.contains(request.getMethod());
+    }
+
+    private boolean refererMatches(HttpServletRequest request) {
+        String referer = request.getHeader("referer");
+        if (referer == null) {
+            return false;
+        }
+        URI uri = URI.create(referer);
+        return uri.getHost().endsWith(".redhat.com") && (uri.getPort() == -1 || uri.getPort() == 443);
+    }
+
+    private boolean originMatches(HttpServletRequest request) {
+        String origin = request.getHeader("origin");
+        return origin != null && (origin.endsWith(".redhat.com") || origin.endsWith(".redhat.com:443"));
```

Comment:
Should we make these configurable so that if we had to change anything like the port, we could tweak it a little more easily? some sort of whitelist?

Reply:
Sure. I will make the domain suffix and port configurable, with defaults. I don't think we need to handle more than one domain, so for the principle of YAGNI (you ain't gonna need it), will leave it at those two.